### ISSUE TITLE
Beatmap listing supporter-only filter show "...nope, nothing found." instead of telling you that you need an active osu!supporter tag

### ISSUE THREAD

**Issue by mouzedrift** [ppy/osu#13530]
**Screenshots or videos showing encountered issue:** 
osu!web:
![image](https://user-images.githubusercontent.com/43358824/122264059-686a0080-ced7-11eb-8a44-2357bd1a2410.png)

osu!lazer:
![image](https://user-images.githubusercontent.com/43358824/122264015-5c7e3e80-ced7-11eb-9883-7c288d66bb6c.png)

**osu!lazer version:** 
2021.612.0

**Comment by JimmyC7834** [comment]
There's no supporter-required.png in the osu-resources/tree/master/osu.Game.Resources/Textures/Online folder.
![supporter-required](https://user-images.githubusercontent.com/42904673/122561856-ed3e5100-d074-11eb-886c-8bcf6ff0db9d.png)

**Comment by bdach** [comment]
If you are interesting in implementing this then it should be fine to PR that addition to resources.

**Comment by JimmyC7834** [comment]
> If you are interesting in implementing this then it should be fine to PR that addition to resources.

Sure, but I'm new and don't know how to do that

**Comment by bdach** [comment]
We don't have the bandwidth to provide individual entry-level advice how to start contributing to open source projects. Please look for online resources on this and preferably also read [the contributing guidelines](https://github.com/ppy/osu/blob/master/CONTRIBUTING.md) if you decide to pursue this further.

**Comment by JimmyC7834** [comment]
Sorry, my bad. I didn't know that PR means pull request

**Comment by JimmyC7834** [comment]
![image](https://user-images.githubusercontent.com/42904673/122634508-21674f80-d111-11eb-8576-1415fa55a7f9.png)
I've managed to make this layout but the font asset failed to load.
![image](https://user-images.githubusercontent.com/42904673/122634538-4eb3fd80-d111-11eb-91e2-76bdc1653c5d.png)

I'm using the LinkFlowContainer with:
```
linkFlowContainer.AddText(
                    text[0],
                    t =>
                    {
                        t.Font = OsuFont.GetFont(size: 16);
                    }
                );
```

**Comment by gagahpangeran** [comment]
Is your branch from latest head in master? (both `osu` and `osu-resource`)

There's some changes in font folder yesterday and one of your local repo, either `osu` or `osu-resource`, might be mismatched.

**Comment by JimmyC7834** [comment]
Yes, you're right. I've updated my branch and the font asset loaded successfully.
Are there any extra requirements for the text style? If not, should I make a PR for the changes and wait for review?

**Comment by bdach** [comment]
The design of the supporter placeholder should match web/use web values exactly wrt component sizings and layout. Please use the "inspect element" functionality all major web browsers include nowadays to see them, unless you prefer to read osu-web source.

Please make sure to also include a test for it as it's not an easily testable part of the system as it depends on the logged-in user.

**Comment by JimmyC7834** [comment]
Sure! Appreciate the guide.
I've tested it out on my account (non-supporter) and it works.
But I don't own a supporter account so I haven't test with one yet.
Also, I should make my own tests in osu.Game.Tests\Visual\SongSelect, right?

**Comment by bdach** [comment]
> But I don't own a supporter account so I haven't test with one yet.

Which is precisely why I'm asking to include a visual test for this, so that you don't have to. There are other test scenes that allow toggling whether a mock/test user has supporter (see [example](https://github.com/ppy/osu/blob/7654df94f6f37b8382be7dfcb4f674e03bd35427/osu.Game.Tests/Visual/Menus/TestSceneDisclaimer.cs#L16-L26)).

> Also, I should make my own tests in osu.Game.Tests\Visual\SongSelect, right?

No. Song select is not the beatmap listing overlay. You want `osu.Game.Tests/Visual/Online`. Maybe even you can use [`TestSceneBeatmapSetOverlay`](https://github.com/ppy/osu/blob/17ad214e0ee6cbe6c7c063e211699536d6fab75d/osu.Game.Tests/Visual/Online/TestSceneBeatmapSetOverlay.cs).

**Comment by gagahpangeran** [comment]
And another information, when you run in Debug mode, it automatically connect to dev server. Register new account in dev server and your account in dev server will automatically has supporter tag. Make sure you run in Debug and not Release to avoid register new account in main server.

**Comment by bdach** [comment]
@gagahpangeran quirks of debug/release configurations should not be relied on to test changes. the visual test browser is there for that.

**Comment by gagahpangeran** [comment]
Yeah, I just give information if they want to test in real environment. Apart from writing real test.

**Comment by JimmyC7834** [comment]
I've make a test for it and this is a part of the test:
``` c#
[Test]
        public void TestSupporterOnlyFiltersPlaceholder() {

            AddStep("toggle non-supporter", () =>
            {
                // non-supporter user
                ((DummyAPIAccess)API).LocalUser.Value = new User
                {
                    Username = API.LocalUser.Value.Username,
                    Id = API.LocalUser.Value.Id + 1,
                    IsSupporter = false,
                };
            });

            AddStep("toggle Random Rank Achieved filter", () => {
                setsForResponse.Clear();
                overlay.ChildrenOfType<BeatmapListingSearchControl>().Single().Ranks.Clear();
                Scoring.ScoreRank r = (Scoring.ScoreRank)(TestContext.CurrentContext.Random.NextShort() % 8);
                overlay.ChildrenOfType<BeatmapListingSearchControl>().Single().Ranks.Add(r);
            });

            AddUntilStep("supporter-placeholder show", () => overlay.ChildrenOfType<BeatmapListingOverlay.SupporterRequiredDrawable>().SingleOrDefault()?.IsPresent == true);

```

It passed the test but I don't know how to visualize the test process.

**Comment by peppy** [comment]
Please read the contributing guide. You should be using visual tests. You should be familiar with the visual test browser before starting to work on this project. It is the basis for anything and everything.

**Comment by JimmyC7834** [comment]
Sorry for being an idiot. I've read through the contributing guide about testing and [Dynamic Compilation and Visual Testing](https://github.com/ppy/osu-framework/wiki/Development-and-Testing) in the wiki. But I'm still not sure how to set up the visual test. Could anyone give me a tip?

**Comment by JimmyC7834** [comment]
> Sorry for being an idiot. I've read through the contributing guide about testing and [Dynamic Compilation and Visual Testing](https://github.com/ppy/osu-framework/wiki/Development-and-Testing) in the wiki. But I'm still not sure how to set up the visual test. Could anyone give me a tip?

Okay, I got the visual test working and the test worked.

**Comment by JimmyC7834** [comment]
I've made the visual test and passed it. Should I include the test detail here?

**Comment by bdach** [comment]
Why would that be required? The test should be a part of any PR that intends to resolve this issue.

**Comment by JimmyC7834** [comment]
> Why would that be required? The test should be a part of any PR that intends to resolve this issue.

Oh, so I would need to create a PR for this issue and discuss the testing details there, right?

**Comment by bdach** [comment]
I'm not sure if you've read the contributing guidelines - as we've asked you no less than twice now - but if you did you would know that PRing changes is the only way of submitting any code changes to this repository.

We cannot continue to provide this level of detailed entry-level advice. We're not a special project in this respect. Submitting changes as pull requests is how 99% of open source software works. Please try to figure out the basics on your own.

**Comment by JimmyC7834** [comment]
@bdach Okay, got it. I apologize for the inconvenience I cause. This is my first open-source project and I'm not confident to a part of such a large project. I would figure out the things on my own from now on. Thanks for the help so far. Please correct me if I do something wrong in the future.

**Comment by JimmyC7834** [comment]
I'm using ```BeatmapsStrings.ListingSearchSupporterFilterQuoteDefault``` to get the supporter required text. It takes two strings but that makes the passed text (ListingSearchFiltersPlayed and ListingSearchFiltersRank) would not be translated in-game.

**Comment by bdach** [comment]
That is going to require support of localisation in `TextFlowContainer` and co., or a mechanism to format `LocalisableString`s using other `LocalisableString`s. I do not recommend you attempt to add this.

**Comment by JimmyC7834** [comment]
I've noticed that all the languages have the osu!supporter tag at the end of the sentence so I could make the supporter required text translated but with the filters text remains in English. I suppose it is not approved so maybe this issue would need to be held wait until someone added the corresponding mechanism.

**Comment by gagahpangeran** [comment]
Just using the string literal for now. The localisation is a separate issue.
